### DISCUSSION TITLE
Force quit does not cancel on-going tasks

### DISCUSSION THREAD

**Discussion by vxgmichel** [Kludex/uvicorn#2525] [Potential Issue]
I've seen several [related issues](https://github.com/encode/uvicorn/discussions/2327), but here's a very simple way to reproduce the fact that the "force quit" feature does not cancel on-going tasks as I expected:

```python
import asyncio
from starlette.applications import Starlette
from starlette.routing import Route


async def homepage(request):
    while True:
        await asyncio.sleep(1)


app = Starlette(debug=True, routes=[Route('/', homepage)])
```

Run the app:
```shell
$ uvicorn main:app
INFO:     Started server process [99431]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
```

Run a client in another term:
```shell
$ curl http://127.0.0.1:8000
```

Try to stop the server with several `ctrl+c` (it won't stop):
```shell
^CINFO:     Shutting down
INFO:     Waiting for connections to close. (CTRL+C to force quit)
^C^C^C^C^C^C^C^C^C^C^C
```

Using `--timeout-graceful-shutdown` does work, but I would still expect to be able to force the cancelling of on-going requests.

Is it a bug or the expected behavior?

**Comment by vxgmichel** [comment]
As far as I can tell, this behavior is related to this code:
https://github.com/encode/uvicorn/blob/b7241e1c64b8e10f56191d2836fa386ad3e56681/uvicorn/server.py#L296-L312

Note how `self.force_exit` skips the waiting for existing tasks and connections, but there is still the `await server.wait_closed()` at the end that will block, waiting for socket transport to close.

The graceful shutdown timeout is not affected by this issue because the whole `_wait_tasks_to_complete` method is wrapped in a `wait_for`:
https://github.com/encode/uvicorn/blob/b7241e1c64b8e10f56191d2836fa386ad3e56681/uvicorn/server.py#L277-L280

**Comment by Kludex** [comment]
I understand your point, but it runs as expected.

**Reply by Kludex** [reply]
And that's why we added the `timeout` graceful flag.

**Reply by vxgmichel** [reply]
But isn't it weird that the graceful shutdown timeout and the forced exit do not behave the same?

My intuition would have been that hitting `ctrl+c` a second time would be the same as shortening the graceful shutdown timeout to "right now".

Plus the `_wait_tasks_to_complete` code seems to points to this direction too: having `force_exit` set to `True` means we're skipping the waiting for existing tasks and connections, but for some reasons we do not skip the `server.wait_closed()` operation.
